e2e/pos_test.go

```diff
@@ -0,0 +1,617 @@
+			Input:    framework.MethodSig(methodName),
+		}, senderKey)
+
+		if signErr != nil {
+			t.Fatalf("Unable to sign transaction, %v", signErr)
+		}
+
+		currentNonce++
+
+		return signedTx
+	}
+
+	oneEth := framework.EthToWei(1)
+	zeroEth := framework.EthToWei(0)
+	for i := 0; i < numTransactions; i++ {
+		var msg *txpoolOp.AddTxnReq
+		if i%2 == 0 {
+			unstakeTxn := generateTx(zeroEth, "unstake")
+			msg = &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: unstakeTxn.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			}
+		} else {
+			stakeTxn := generateTx(oneEth, "stake")
+			msg = &txpoolOp.AddTxnReq{
+				Raw: &any.Any{
+					Value: stakeTxn.MarshalRLP(),
+				},
+				From: types.ZeroAddress.String(),
+			}
+		}
+
+		_, addErr := txpoolClient.AddTxn(context.Background(), msg)
+		if addErr != nil {
+			t.Fatalf("Unable to add txn, %v", addErr)
+		}
+	}
+
+	// Set up the blockchain listener to catch the added block event
+	blockNum := waitForBlock(t, srv, 1, 0)
+
+	block, blockErr := client.Eth().GetBlockByNumber(web3.BlockNumber(blockNum), true)
+	if blockErr != nil {
+		t.Fatalf("Unable to fetch block")
+	}
+
+	// Find how much the account paid for all the transactions in this block
+	paidFee := big.NewInt(0).Mul(bigGasPrice, big.NewInt(int64(block.GasUsed)))
+
+	// Check the balances
+	actualAccountBalance := framework.GetAccountBalance(senderAddr, client, t)
+	actualStakingSCBalance, fetchError := framework.GetStakedAmount(senderAddr, client)
+	if fetchError != nil {
+		t.Fatalf("Unable to fetch staking SC balance, %v", fetchError)
+	}
+
+	expStake := big.NewInt(0).Mul(big.NewInt(int64(numDummyStakers)), bigDefaultStakedBalance)
+	expStake.Add(expStake, oneEth)
+
+	assert.Equalf(t,
+		expStake.String(),
+		actualStakingSCBalance.String(),
+		"Staked address balance mismatch after stake / unstake exploit",
+	)
+
+	// Make sure the account balances match up
+
+	// expBalance = previousAccountBalance + stakeRefund - 1 ETH - block fees
+	expBalance := big.NewInt(0).Sub(big.NewInt(0).Add(defaultBalance, bigDefaultStakedBalance), oneEth)
+	expBalance = big.NewInt(0).Sub(expBalance, paidFee)
+
+	assert.Equalf(t,
+		expBalance.String(),
+		actualAccountBalance.String(),
+		"Account balance mismatch after stake / unstake exploit",
+	)
+}
+
+// Test scenario:
+// User has 0 ETH staked and a balance of 100 ETH
+// Stake 2 ETH -> Unstake
+// Expected result for tests: Staked: 0 ETH; Balance: ~100 ETH
+func TestPoS_StakeUnstakeWithinSameBlock(t *testing.T) {
+	// Predefined values
+	var blockGasLimit uint64 = 5000000000
+	stakingContractAddr := staking.AddrStakingContract
+	defaultBalance := framework.EthToWei(100)
+	bigGasPrice := big.NewInt(framework.DefaultGasPrice)
+
+	senderKey, senderAddr := tests.GenerateKeyAndAddr(t)
+	numDummyStakers := 10
+
+	devInterval := 5 // s
+
+	// Set up the test server
+	srvs := framework.NewTestServers(t, 1, func(config *framework.TestServerConfig) {
```

Comment:
More validators please so that the initial staking amount doesn't only consist of this user's tokens.

Reply:
This test has `11` stakers:
https://github.com/0xPolygon/polygon-sdk/blob/eb35d49a64ab3c8d044e3333c09e68cab7a6443e/e2e/pos_test.go#L516